### ISSUE TITLE
The implementation of `cond_fn` does not match the original paper (Diffusion Models Beat GANs on Image Synthesis).

### ISSUE THREAD

**Issue by yuxi-liu-wired** [Jack000/glid-3-xl#15]
Specifically, this line in [`sample.py`](https://github.com/Jack000/glid-3-xl/blob/master/sample.py): `x_in = out['pred_xstart'] * fac + x * (1 - fac)`. The full code is copied for reference:

```python
    def cond_fn(x, t, context=None, clip_embed=None, image_embed=None):
        with torch.enable_grad():
            x = x[:args.batch_size].detach().requires_grad_()

            n = x.shape[0]

            my_t = torch.ones([n], device=device, dtype=torch.long) * cur_t

            kw = {
                'context': context[:args.batch_size],
                'clip_embed': clip_embed[:args.batch_size] if model_params['clip_embed_dim'] else None,
                'image_embed': image_embed[:args.batch_size] if image_embed is not None else None
            }

            out = diffusion.p_mean_variance(model, x, my_t, clip_denoised=False, model_kwargs=kw)

            fac = diffusion.sqrt_one_minus_alphas_cumprod[cur_t]
            x_in = out['pred_xstart'] * fac + x * (1 - fac)

            x_in /= 0.18215

            x_img = ldm.decode(x_in)

            clip_in = normalize(make_cutouts(x_img.add(1).div(2)))
            clip_embeds = clip_model.encode_image(clip_in).float()
            dists = spherical_dist_loss(clip_embeds.unsqueeze(1), text_emb_clip.unsqueeze(0))
            dists = dists.view([args.cutn, n, -1])

            losses = dists.sum(2).mean(0)

            loss = losses.sum() * args.clip_guidance_scale

            return -torch.autograd.grad(loss, x)[0]
 ```

As far as I can see, this line in the original paper's notation would be written as $x_{in} = \sqrt{1 - \bar \alpha_t} x_0' + (1 - \sqrt{1 - \bar \alpha_t})x_t$. However, this does not correspond to *any* formula in the original paper. I think the code can be made exactly equal to the original paper if we replace that line with `x_in = x`, since the original paper uses the loss function defined as $L(x_t) = - \ln p_\phi(y | x_t)$, and $x_t$ is here implemented as `x`.

In fact, in the [original OpenAI github implementation](https://github.com/openai/guided-diffusion/blob/main/scripts/classifier_sample.py), we do indeed see `x_in = x.detach().requires_grad_(True)`, exactly the same as how it is implemented in the paper.

```python
    def cond_fn(x, t, y=None):
        assert y is not None
        with th.enable_grad():
            x_in = x.detach().requires_grad_(True)
            logits = classifier(x_in, t)
            log_probs = F.log_softmax(logits, dim=-1)
            selected = log_probs[range(len(logits)), y.view(-1)]
            return th.autograd.grad(selected.sum(), x_in)[0] * args.classifier_scale
```
